selfhost/typechecker.jakt

```diff
@@ -4463,77 +4463,8 @@ struct Typechecker {
-                                struct_id = match .get_type(args[0]) {
-                                    Struct(struct_id) | GenericInstance(id: struct_id) => StructOrEnumId::Struct(struct_id)
-                                    Enum(id) | GenericEnumInstance(id) => StructOrEnumId::Enum(id)
-                                    else => {
-                                        .error("Can't use non-struct type as an optional type in optional chained call", span)
-                                        found_optional = false
-                                        yield StructOrEnumId::Struct(optional_struct_id)
-                                    }
-                                }
-                            }
-
-                            yield Some(struct_id ?? StructOrEnumId::Struct(optional_struct_id))
-                        }
-                        else => Some(StructOrEnumId::Struct(id))
-                    }
-                }
-                GenericEnumInstance(id) => Some(StructOrEnumId::Enum(id))
-                else => {
-                    .error(message: format("no methods available on value (type: {})", .type_name(type_id: checked_expr_type_id)), span: checked_expr.span())
-                    let none: StructOrEnumId? = None
-
-                    yield none
-                }
-            }
-
-            if is_optional and not found_optional {
-                .error(message: format("Optional chain mismatch: expected optional chain, found {}", .type_name(type_id: checked_expr_type_id)), span: checked_expr.span())
-            }
-
-            let checked_call_expr = .typecheck_call(call, caller_scope_id: scope_id, span, this_expr: checked_expr, parent_id, safety_mode, type_hint, must_be_enum_constructor: false)
-            let type_id = checked_call_expr.type()
-            yield match checked_call_expr {
-                Call(call) => {
-                    mut result_type = call.return_type
-                    if is_optional {
-                        let optional_struct_id = .find_struct_in_prelude("Optional")
-                        result_type = .find_or_add_type_id(Type::GenericInstance(id: optional_struct_id, args: [result_type]))
-                    }
-                    yield CheckedExpression::MethodCall(
-                        expr: checked_expr
-                        call
-                        span
-                        is_optional
-                        type_id: result_type)
-                }
-                else => {
-                    .compiler.panic("typecheck_call should return `CheckedExpression::Call()`")
-                }
-            }
-        }
+        Call(call, span) => .typecheck_call_expr(call, span, must_be_enum_constructor: false, scope_id, safety_mode, type_hint)
+        MethodCall(expr, call, span, is_optional) => .typecheck_method_call_expr(expr, call, span, is_optional, scope_id, safety_mode, type_hint)
```

Comment:
is "call" supposed to be a superset of "method_call"? from the naming I would guess that, but given they are called from distinct enum cases contradicts that

Reply:
No, "call" is a qualified call (e.g `foo()`), methodcall is the one with a dot `foo.bar()`. They have different ways of resolving, hence this refactor.